app/src/ui/keyboardlayoutprovider.cpp

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright 2019 Lieven Hey <t.schilling@snafu.de>.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "keyboardlayoutprovider.h"
+#include <QFile>
+#include <QStandardPaths>
+#include <QJsonDocument>
+#include <QJsonArray>
+#include <QDebug>
+#include <sailfishapp.h>
+#include <QtFeedback/QFeedbackEffect>
+#include <QRegularExpression>
+
+KeyboardLayoutProvider::KeyboardLayoutProvider(QObject *parent) : QObject(parent)
+{
+    m_settings.beginGroup(QStringLiteral("RemoteKeyboard"));
+    m_layout = m_settings.value("layout", "en").toString();
+    m_repeatInterval = m_settings.value("repeatInterval", 200).toInt();
+    m_feedback = m_settings.value("keyboardFeedback", true).toBool();
+    m_settings.endGroup();
+    setLayout(m_layout);
+}
+
+QString KeyboardLayoutProvider::layout() const
+{
+    return m_layout;
+}
+
+void KeyboardLayoutProvider::setLayout(const QString &layout)
+{
+    // seems to be the only way to get the files stored
+    QFile layoutFile("/usr/share/maliit/plugins/com/jolla/layouts/" + layout + ".qml");
+    if (!layoutFile.open(QIODevice::ReadOnly)) {
+        qDebug() << "unkown layout: " << layout;
```

Comment:
Typo and please use an own logger for KeyboardLayoutProvider (see other files in app/src/ui).
```suggestion
        qCDebug(logger) << "Unknown layout:" << layout;
```